crates/unftp-sbe-gcs/src/lib.rs

```diff
@@ -134,30 +137,93 @@ impl CloudStorage {
 
-    // TODO: Cache the token. For `ServiceAccountKey`, the oauth client would already cache the token - we just need to move it to `CloudStorage`. For `WorkloadIdentity`, we can cache it in `CloudStorage`.
     #[tracing_attributes::instrument]
-    async fn get_token(&self) -> Result<String, Error> {
+    async fn get_token_value(&self) -> Result<String, Error> {
+        {
+            let cache = self.cached_token.read().unwrap();
+            if let Some(token) = &*cache {
+                if token.active() {
+                    return Ok(token.value.clone());
+                }
+            }
+        }
+
+        let token = self.fetch_token().await?;
+        {
+            let mut cache = self.cached_token.write().unwrap();
+            let cached = cache.insert(token);
+
+            Ok(cached.value.clone())
+        }
+    }
+
+    async fn fetch_token(&self) -> Result<Token, Error> {
         match &self.auth {
             AuthMethod::ServiceAccountKey(_) => {
                 let key = self.auth.to_service_account_key()?;
                 let auth = ServiceAccountAuthenticator::builder(key).hyper_client(self.client.clone()).build().await?;
 
                 auth.token(&["https://www.googleapis.com/auth/devstorage.read_write"])
-                    .map_ok(|t| t.as_str().to_string())
+                    .map_ok(|t| t.into())
                     .map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))
                     .await
             }
-            AuthMethod::WorkloadIdentity(service) => workflow_identity::request_token(service.clone(), self.client.clone())
-                .await
-                .map(|t| t.access_token),
-            AuthMethod::None => Ok("unftp_test".to_string()),
+            AuthMethod::WorkloadIdentity(service) => workload_identity::request_token(service.clone(), self.client.clone()).await.map(|t| t.into()),
+            AuthMethod::None => Ok(Token {
+                value: "unftp_test".to_string(),
+                expires_at: None,
+            }),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Token {
+    value: String,
+    expires_at: Option<time::OffsetDateTime>,
+}
+
+impl Token {
+    /// active yields true when the token is present and has not expired. In all other cases, it
+    /// returns false.
+    fn active(&self) -> bool {
+        self.expires_at
+            .map(|expires_at| {
+                let now = time::OffsetDateTime::now_utc();
+                let safety_margin = time::Duration::seconds(5);
+
+                expires_at > (now - safety_margin)
+            })
+            .unwrap_or(false)
+    }
+}
+
+impl From<yup_oauth2::AccessToken> for Token {
+    fn from(yup: yup_oauth2::AccessToken) -> Self {
+        Self {
+            value: yup.as_str().to_string(),
+            expires_at: yup.expiration_time(),
+        }
+    }
+}
+
+impl From<workload_identity::TokenResponse> for Token {
+    fn from(wfi: workload_identity::TokenResponse) -> Self {
```

Comment:
wli instead of wfi or what does wfi stand for?

Reply:
Hah, I named that argument before I changed the name from workflow to workload :D

Reply:
I changed the name to from_token in both the workload_identity and service_account implementations. That name should be a bit more descriptive.

Reply:
No need to change it but even from_token is a bit ambiguous since you're also converting `to` a token.  :-) . `source` or something like that?

Reply:
Agreed and done